### PR TITLE
cacher: Fix watch behaviour for unset RV

### LINKED ISSUE #114845: Watchcache doesn't distinguish the RV=0 and RV="" semantic for watches
As Jordan pointed out in:
https://github.com/kubernetes/enhancements/pull/3667/files#r1055906242

watchcache treats RV=0 and RV="" the same way, even though it's not what we document:
https://kubernetes.io/docs/reference/using-api/api-concepts/#semantics-for-watch

This should be fixed.

/kind bug
/sig api-machinery
/priority important-soon

### PR DESCRIPTION
**Pull request by MadhavJivrajani**
#### What type of PR is this?

/kind bug cleanup
/priority important-soon
#### What this PR does / why we need it:

This PR consists of 2 commits:
- #### Commit 1 (prep for commit 2):
  - This commit allows injecting errors for the Watch() method of the dummy storage impl.
  - As a consequence of this, a [race](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/115096/pull-kubernetes-unit/1614863523177828352) is introduced between when the injected error is written and read whenever a Watch() is invoked using the dummy storage. This commit adds locking in order to mitigate this.

- #### Commit 2:
  - The original design was to honour strong consistency semantics for when the RV is unset, i.e. serve the watch by
doing a quorum read.
  - However, the implementation did not match the intent, in that, the Cacher did not distinguish between set and unset RV.
This commit rectifies that behaviour by serving the watch from the underlying storage if the RV is unset.
  - This commit subsequently also adds a test for the same.

#### Which issue(s) this PR fixes:

Fixes #114845

#### Does this PR introduce a user-facing change?

```release-note
cacher: If RV is unset, the watch is now served from the underlying storage as documented.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

/assign @wojtek-t

### CODE REVIEW

**Comment by MadhavJivrajani** [comment]
/retest
Related flake reported here: https://github.com/kubernetes/kubernetes/issues/114852

**Comment by MadhavJivrajani** [comment]
/retest
Flake: https://github.com/kubernetes/kubernetes/issues/106579

**Comment by MadhavJivrajani** [comment]
@wojtek-t friendly ping for review

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 1]
```diff
@@ -482,6 +482,13 @@ func (c *Cacher) Delete(
 // Watch implements storage.Interface.
 func (c *Cacher) Watch(ctx context.Context, key string, opts storage.ListOptions) (watch.Interface, error) {
 	pred := opts.Predicate
+	// If the resourceVersion is unset, ensure that the rv
+	// from which the watch is being served, is the latest
+	// one. "latest" is ensured by serving the watch from
+	// the underlying storage.
+	if opts.ResourceVersion == "" {
+		return c.storage.Watch(ctx, key, opts)
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:490]: This is technically much more expensive request, but given there should be very few of those and eventually we can optimize it, I'm fine with it for now.

**Comment by wojtek-t** [comment]
/lgtm
/approve

Thanks!

**Comment by MadhavJivrajani** [comment]
@wojtek-t considering this was a bug, let's open cherry picks?

**Comment by wojtek-t** [comment]
I'm not sure - it's subtle enough that I wouldn't do that now.

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 1 above)
Comment (MadhavJivrajani) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go:490]: Gotcha!
When you say optimise it, do you mean in terms of further reducing the number of times we actually have to call into etcd by changing the if condition?

**Pull request merged**